internal/settings/providers/inwx/provider.go

```diff
@@ -0,0 +1,140 @@
+package inwx
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/qdm12/ddns-updater/internal/models"
+	"github.com/qdm12/ddns-updater/internal/settings/errors"
+	"github.com/qdm12/ddns-updater/internal/settings/headers"
+	"github.com/qdm12/ddns-updater/internal/settings/utils"
+	"github.com/qdm12/ddns-updater/pkg/publicip/ipversion"
+)
+
+type Provider struct {
+	domain    string
+	host      string
+	ipVersion ipversion.IPVersion
+	username  string
+	password  string
+}
+
+func New(data json.RawMessage, domain, host string,
+	ipVersion ipversion.IPVersion) (p *Provider, err error) {
+	extraSettings := struct {
+		Username string `json:"username"`
+		Password string `json:"password"`
+	}{}
+	if err := json.Unmarshal(data, &extraSettings); err != nil {
+		return nil, err
+	}
+	p = &Provider{
+		domain:    domain,
+		host:      host,
+		ipVersion: ipVersion,
+		username:  extraSettings.Username,
+		password:  extraSettings.Password,
+	}
+	if err := p.isValid(); err != nil {
+		return nil, err
+	}
+	return p, nil
+}
+
+func (p *Provider) isValid() error {
+	switch {
+	case len(p.username) == 0:
+		return errors.ErrEmptyUsername
+	case len(p.password) == 0:
+		return errors.ErrEmptyPassword
+	case p.host == "*":
+		return errors.ErrHostWildcard
```

Comment:
My bad, some of my own old code is not the greatest, I need to update it to better standards. Here are two suggestions:

- Compare strings with `== ""` instead of a length check (less weird)
- Wrap errors with `fmt.Errorf` so context can be added later and callers are forced to use `errors.Is(err, errors.ErrEmptyUsername)` and cannot use `err == errors.ErrEmptyUsername`. A bit of a long topic to explain, more in https://go.dev/blog/go1.13-errors

```suggestion
	case p.username == "":
		return fmt.Errorf("%w", errors.ErrEmptyUsername)
	case p.password == "":
		return fmt.Errorf("%w", errors.ErrEmptyPassword)
	case p.host == "*":
		return fmt.Errorf("%w", errors.ErrHostWildcard)
```